compiler/noirc_frontend/src/elaborator/lints.rs

```diff
@@ -0,0 +1,221 @@
+    } else {
+        None
+    }
+}
+
+/// Attempting to define new low level (`#[builtin]` or `#[foreign]`) functions outside of the stdlib is disallowed.
+pub(super) fn low_level_function_outside_stdlib(
+    func: &NoirFunction,
+    crate_id: CrateId,
+) -> Option<ResolverError> {
+    let is_low_level_function =
+        func.attributes().function.as_ref().map_or(false, |func| func.is_low_level());
+    if !crate_id.is_stdlib() && is_low_level_function {
+        Some(ResolverError::LowLevelFunctionOutsideOfStdlib { ident: func.name_ident().clone() })
+    } else {
+        None
+    }
+}
+
+/// `pub` is required on return types for entry point functions
+pub(super) fn missing_pub(func: &NoirFunction, is_entry_point: bool) -> Option<ResolverError> {
+    if is_entry_point
+        && func.return_type().typ != UnresolvedTypeData::Unit
+        && func.def.return_visibility == Visibility::Private
+    {
+        Some(ResolverError::NecessaryPub { ident: func.name_ident().clone() })
+    } else {
+        None
+    }
+}
+
+/// `#[recursive]` attribute is only allowed for entry point functions
+pub(super) fn recursive_non_entrypoint_function(
+    func: &NoirFunction,
+    is_entry_point: bool,
+) -> Option<ResolverError> {
+    if !is_entry_point && func.kind == FunctionKind::Recursive {
+        Some(ResolverError::MisplacedRecursiveAttribute { ident: func.name_ident().clone() })
+    } else {
+        None
+    }
+}
+
+/// Test functions cannot have arguments in order to be executable.
+pub(super) fn test_function_with_args(func: &NoirFunction) -> Option<ResolverError> {
+    if func.attributes().is_test_function() && !func.parameters().is_empty() {
+        Some(ResolverError::TestFunctionHasParameters { span: func.name_ident().span() })
+    } else {
+        None
+    }
+}
+
+/// Check that we are not passing a mutable reference from a constrained runtime to an unconstrained runtime.
+pub(super) fn unconstrained_function_args(
+    function_args: &[(Type, ExprId, Span)],
+) -> Vec<TypeCheckError> {
+    function_args
+        .iter()
+        .filter_map(|(typ, _, span)| {
+            if type_contains_mutable_reference(typ) {
+                Some(TypeCheckError::ConstrainedReferenceToUnconstrained { span: *span })
+            } else {
+                None
+            }
+        })
+        .collect()
+}
+
+/// Check that we are not passing a slice from an unconstrained runtime to a constrained runtime.
+pub(super) fn unconstrained_function_return(
+    return_type: &Type,
+    span: Span,
+) -> Option<TypeCheckError> {
+    if return_type.contains_slice() {
+        Some(TypeCheckError::UnconstrainedSliceReturnToConstrained { span })
+    } else if type_contains_mutable_reference(return_type) {
+        Some(TypeCheckError::UnconstrainedReferenceToConstrained { span })
+    } else {
+        None
+    }
+}
+
+fn type_contains_mutable_reference(typ: &Type) -> bool {
+    matches!(&typ.follow_bindings(), Type::MutableReference(_))
+}
```

Comment:
I was going to suggest fixing this function but I suppose that is what https://github.com/noir-lang/noir/pull/5009 is for